### PR TITLE
Update MVDR beamforming tutorial

### PR DESCRIPTION
**Pull request by nateanl**
- Use `download_asset` to download audios.
- Replace `MVDR` module with new-added `SoudenMVDR` and `RTFMVDR` modules.
- Benchmark performances of `F.rtf_evd` and `F.rtf_power` for RTF computation.
- Visualize the spectrograms and masks.

### CODE REVIEW

**Comment by nateanl** [comment]
tutorial: https://output.circle-artifacts.com/output/job/de0ccc46-2998-42fe-a2fa-7ac1c24f4205/artifacts/0/docs/tutorials/mvdr_tutorial.html

File: examples/tutorials/mvdr_tutorial.py [hunk 1]
```diff
@@ -1,32 +1,46 @@
 """
-MVDR with torchaudio
-====================
+MVDR Beamforming with torchaudio
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:2]:
```suggestion
MVDR Beamforming with TorchAudio
```

File: examples/tutorials/mvdr_tutorial.py [hunk 2]
```diff
@@ -1,32 +1,46 @@
...
-# Overview
-# --------
+# 1. Overview
+# -----------
 #
-# This is a tutorial on how to apply MVDR beamforming with
-# :py:func:`torchaudio.transforms.MVDR`.
+# This is a tutorial on how to apply Minimum Variance Distortionless
+# Response (MVDR) beamforming to estimate the enhanced speech with
+# torchaudio.
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:16]:
```suggestion
# This is a tutorial on applying Minimum Variance Distortionless
# Response (MVDR) beamforming to estimate enhanced speech with
# TorchAudio.
```

File: examples/tutorials/mvdr_tutorial.py [hunk 3]
```diff
@@ -1,32 +1,46 @@
...
+#    magnitude by the mixture magnitude.
+# -  Power spectral density (PSD) matrices are estimated by
+#    :py:func:`torchaudio.transforms.PSD`.
+# -  The enhanced speech is estimated by using the two MVDR modules in
+#    torchaudio (:py:func:`torchaudio.transforms.SoudenMVDR` and
+#    :py:func:`torchaudio.transforms.RTFMVDR`).
+# -  We benchmark the two methods
+#    (:py:func:`torchaudio.functional.rtf_evd` and
+#    :py:func:`torchaudio.functional.rtf_power`) for computing the
+#    relative transfer function (RTF) matrix of the reference microphone.
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:30]: i would use active voice for these and maintain consistent structure, e.g.

```suggestion
# -  Generate an ideal ratio mask (IRM) by dividing the clean/noise
#    magnitude by the mixture magnitude.
# -  Estimate power spectral density (PSD) matrices using :py:func:`torchaudio.transforms.PSD`.
# -  Estimate enhanced speech using MVDR modules 
#    :py:func:`torchaudio.transforms.SoudenMVDR` and
#    :py:func:`torchaudio.transforms.RTFMVDR`).
# -  Benchmark the two methods
#    (:py:func:`torchaudio.functional.rtf_evd` and
#    :py:func:`torchaudio.functional.rtf_power`) for computing the
#    relative transfer function (RTF) matrix of the reference microphone.
```

File: examples/tutorials/mvdr_tutorial.py [hunk 4]
```diff
@@ -1,32 +1,46 @@
...
+# 1. Overview
+# -----------
 #
-# This is a tutorial on how to apply MVDR beamforming with
-# :py:func:`torchaudio.transforms.MVDR`.
+# This is a tutorial on how to apply Minimum Variance Distortionless
+# Response (MVDR) beamforming to estimate the enhanced speech with
+# torchaudio.
 #
 # Steps
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:18]:
```suggestion
# Steps:
```

File: examples/tutorials/mvdr_tutorial.py [hunk 5]
```diff
@@ -38,258 +52,315 @@
 #
 #    ``SSB07200001\#noise-sound-bible-0038\#7.86_6.16_3.00_3.14_4.84_134.5285_191.7899_0.4735\#15217\#25.16333303751458\#0.2101221178590021.wav``
 #
-# which was generated with;
+# which was generated with:
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:55]:
```suggestion
# , which was generated with:
```

File: examples/tutorials/mvdr_tutorial.py [hunk 6]
```diff
@@ -1,32 +1,46 @@
...
 
 **Author** `Zhaoheng Ni <zni@fb.com>`__
 
 """
 
+
 ######################################################################
-# Overview
-# --------
+# 1. Overview
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:11]: is there a way to generate section numbering automatically? would be nice to not have to worry about it and to be able to easily apply it to all tutorials

File: examples/tutorials/mvdr_tutorial.py [hunk 7]
```diff
@@ -38,258 +52,315 @@
...
-reverb_clean = reverb_clean.to(torch.double)
+waveform_mix = waveform_mix.to(torch.double)
+waveform_clean = waveform_clean.to(torch.double)
+waveform_noise = waveform_noise.to(torch.double)
+
 
 ######################################################################
-# Compute STFT
-# ~~~~~~~~~~~~
+# 3.2. Compute complex-valued Spectrums
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:149]:
```suggestion
# 3.2. Compute complex-valued spectrograms
```

File: examples/tutorials/mvdr_tutorial.py [hunk 8]
```diff
@@ -38,258 +52,315 @@
...
+waveform_clean, sr2 = torchaudio.load(SAMPLE_CLEAN)
+assert sr == sr2 == SAMPLE_RATE
+# The mixture waveform is a combination of clean and noise waveforms
+waveform_noise = waveform_mix - waveform_clean
 
-noise = mix - reverb_clean
 
 ######################################################################
+# Note: To imrove the robustness of computation, it is recommended to use the
+# double precision (``torch.float64`` or ``torch.double`` for the waveforms.
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:140]:
```suggestion
# Note: To improve computational robustness, it is recommended to represent
# the waveforms as double-precision floating point (``torch.float64`` or ``torch.double``) values.
```

File: examples/tutorials/mvdr_tutorial.py [hunk 9]
```diff
@@ -38,258 +52,315 @@
...
+# -------------------------------
 #
 
-ipd.Audio(mix[0], rate=16000)
 
 ######################################################################
-# Noise
-# ~~~~~
+# 4.1. Compute PSD matrices
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:248]:
```suggestion
# 4.1. Compute PSD matrices
# ~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/tutorials/mvdr_tutorial.py [hunk 10]
```diff
@@ -38,258 +52,315 @@
...
+
+psd_speech = psd_transform(spectrum_mix, irm_speech)
+psd_noise = psd_transform(spectrum_mix, irm_noise)
 
-ipd.Audio(noise[0], rate=16000)
 
 ######################################################################
-# Clean speech
-# ~~~~~~~~~~~~
+# 4.2. Apply Beamforming
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:262]:
```suggestion
# 4.2. Apply beamforming
```

File: examples/tutorials/mvdr_tutorial.py [hunk 11]
```diff
@@ -38,258 +52,315 @@
...
-ipd.Audio(mix[0], rate=16000)
 
 ######################################################################
-# Noise
-# ~~~~~
+# 4.1. Compute PSD matrices
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 #
+# :py:func:`torchaudio.transforms.PSD` computes the time-invariant PSD matrix given
+# the multi-channel complex-valued spectrum and the Time-Frequency mask.
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:251]:
```suggestion
# the multi-channel complex-valued spectrogram and the time-frequency mask.
```

File: examples/tutorials/mvdr_tutorial.py [hunk 12]
```diff
@@ -38,258 +52,315 @@
...
+    n_fft=N_FFT,
+    hop_length=N_HOP,
     power=None,
 )
-istft = torchaudio.transforms.InverseSpectrogram(n_fft=1024, hop_length=256)
+istft = torchaudio.transforms.InverseSpectrogram(n_fft=N_FFT, hop_length=N_HOP)
+
+spectrum_mix = stft(waveform_mix)
+spectrum_clean = stft(waveform_clean)
+spectrum_noise = stft(waveform_noise)
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:164]: perhaps rename `spectrum` to `spectrogram` or `spectra`? each result represents a time sequence of spectra, so using the singular "spectrum" is a little misleading

File: examples/tutorials/mvdr_tutorial.py [hunk 13]
```diff
@@ -38,258 +52,315 @@
...
-    results_multi[solution] = est
+plot_specgram(spectrum_noise[0], "Spectrogram of Noise (dB)")
+Audio(waveform_noise[0], rate=SAMPLE_RATE)
+
 
 ######################################################################
-# Apply MVDR beamforming by using single-channel masks
-# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# 3.3. Define the reference microphone
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:196]:
```suggestion
# 3.3. Define the reference microphone
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/tutorials/mvdr_tutorial.py [hunk 14]
```diff
@@ -38,258 +52,315 @@
...
+#
+# We choose the first microphone in the array as the reference channel for demonstration.
+# The selection of the reference channel may depend on the design of the microphone array.
 #
-# We use the 1st channel as an example.
-# The channel selection may depend on the design of the microphone array
+# You can also apply a neural network to estimate the reference channel and
+# pass it to the MVDR module in an end-to-end speech enhancement model.
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:202]: i feel like this could be clearer. are we suggesting that such an "end-to-end speech enhancement model" would comprise a neural network that estimates the reference channel plus the MVDR module? or is only the MVDR module part of the end-to-end speech enhancement model?

File: examples/tutorials/mvdr_tutorial.py [hunk 15]
```diff
@@ -38,258 +52,315 @@
...
-    mvdr = torchaudio.transforms.MVDR(ref_channel=0, solution=solution, multi_mask=False)
-    stft_est = mvdr(spec_mix, irm_speech[0], irm_noise[0])
-    est = istft(stft_est, length=mix.shape[-1])
-    results_single[solution] = est
 
 ######################################################################
-# Compute Si-SDR scores
-# ~~~~~~~~~~~~~~~~~~~~~
+# 3.4. Compute IRMs for target speech and noise
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:209]:
```suggestion
# 3.4. Compute IRMs for target speech and noise
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/tutorials/mvdr_tutorial.py [hunk 16]
```diff
@@ -38,258 +52,315 @@
...
-ipd.Audio(noise[0], rate=16000)
 
 ######################################################################
-# Clean speech
-# ~~~~~~~~~~~~
+# 4.2. Apply Beamforming
+# ~~~~~~~~~~~~~~~~~~~~~~
+#
+# :py:func:`torchaudio.transforms.SoudenMVDR` takes the multi-channel
+# complexed-valued spectrum of the mixture speech, PSD matrices of
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:266]: in general, throughout the tutorial, we should be more precise with distinguishing between "spectrum" and "spectrogram"

```suggestion
# complex-valued spectrogram of the mixture speech, PSD matrices of
```

File: examples/tutorials/mvdr_tutorial.py [hunk 17]
```diff
@@ -38,258 +52,315 @@
...
+# target speech and noise, and the reference channel as the inputs.
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:267]:
```suggestion
# target speech and noise, and the reference channel inputs.
```

File: examples/tutorials/mvdr_tutorial.py [hunk 18]
```diff
@@ -38,258 +52,315 @@
...
 #
+# The output is a single-channel complex-valued spectrum of the enhanced speech.
+# Then we can obtain the enhanced wavefrom by passing it to the
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:270]:
```suggestion
# We can then obtain the enhanced waveform by passing this output to the
```

File: examples/tutorials/mvdr_tutorial.py [hunk 19]
```diff
@@ -38,258 +52,315 @@
...
+# 5. Beamforming using RTFMVDR
+# ----------------------------
 #
 
-ipd.Audio(results_multi["ref_channel"], rate=16000)
 
 ######################################################################
-# Multi-channel mask, stv_evd solution
-# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# 5.1. Compute PSD matrices
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:296]: to maintain parallelism with the steps you outlined in the beginning, where you identify PSD matrix generation as a distinct step, i'd break this (and the corresponding section for `SoudenMVDR`) out into its own section that immediately follows the ideal ratio mask generation section

File: examples/tutorials/mvdr_tutorial.py [hunk 20]
```diff
@@ -38,258 +52,315 @@
...
+psd_noise = psd_transform(spectrum_mix, irm_noise)
+
 
 ######################################################################
-# Multi-channel mask, stv_power solution
-# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# 5.2. Compute RTF
+# ~~~~~~~~~~~~~~~~
+#
+# There are two methods in torchaudio to compute the RTF matrix of the
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:310]:
```suggestion
# TorchAudio offers two methods for computing the RTF matrix of a
```

File: examples/tutorials/mvdr_tutorial.py [hunk 21]
```diff
@@ -38,258 +52,315 @@
...
+# target speech:
+#
+#    :py:func:`torchaudio.functional.rtf_evd`, which applies eigenvalue
+#    decomposition to the PSD matrix of target speech to get the RTF matrix.
+#
+#    :py:func:`torchaudio.functional.rtf_power`, which applies the power iteration
+#    method. You can tune the number of iterations by changing ``n_iter`` argument.
```
Comment (hwangjeff) [inline examples/tutorials/mvdr_tutorial.py:317]:
```suggestion
#    method. You can specify the number of iterations with argument ``n_iter``.
```

File: examples/tutorials/mvdr_tutorial.py (hunk 16 above)
Comment (hwangjeff) [reply examples/tutorials/mvdr_tutorial.py:266]: all of the feedback in this section also applies to "5.3. Apply Beamforming"

File: examples/tutorials/mvdr_tutorial.py (hunk 6 above)
Comment (nateanl) [reply examples/tutorials/mvdr_tutorial.py:11]: I think so. In our contributing.md file it suggests using https://gist.github.com/chsasank/7218ca16f8d022e02a9c0deb94a310fe to convert the ipynb notebook to sphinx format. We can extend it to support auto-indexing.

File: examples/tutorials/mvdr_tutorial.py [hunk 22]
```diff
@@ -1,32 +1,45 @@
 """
-MVDR with torchaudio
-====================
+MVDR Beamforming with TorchAudio
```
Comment (mthrok) [inline examples/tutorials/mvdr_tutorial.py:2]: I have been feeling `with TorchAudio` is redundant.
Can we rephrase it to `<Task> with MVDR Beamforming`?

From the content of the tutorial, I think `Noise reduction with MVDR Beamforming` is a good title.

File: examples/tutorials/mvdr_tutorial.py (hunk 6 above)
Comment (mthrok) [reply examples/tutorials/mvdr_tutorial.py:11]: Tutorials are not necessarily authored with IPython Notebook. (I directly write these from editor) so that's not a viable option. Would be nice to have auto numbering but sphinx-gallery only does the rendering of source, so I do not think there is a good way to automate it.

Comment (mthrok) [reply examples/tutorials/mvdr_tutorial.py:11]: I started numbering to make sections and subsections flat because subsection was hidden by default. But now that is resolved by #2397, I am okay with removing numbering entirely or at least from subsections.

File: examples/tutorials/mvdr_tutorial.py [hunk 23]
```diff
@@ -38,258 +51,304 @@
...
 
-mix, sr = torchaudio.load("_assets/mix.wav")
-reverb_clean, sr2 = torchaudio.load("_assets/reverb_clean.wav")
-clean, sr3 = torchaudio.load("_assets/clean.wav")
-assert sr == sr2
+waveform_mix, sr = torchaudio.load(SAMPLE_MIX)
+waveform_clean, sr2 = torchaudio.load(SAMPLE_CLEAN)
+assert sr == sr2 == SAMPLE_RATE
+# The mixture waveform is a combination of clean and noise waveforms
+waveform_noise = waveform_mix - waveform_clean
```
Comment (mthrok) [inline examples/tutorials/mvdr_tutorial.py:134]: This part felt unintuitive to me. 
If it were `waveform_mix = waveform_clean + waveform_noise` then it makes more sense.
Can we create another asset and fetch it in the same way as the rest?

File: examples/tutorials/mvdr_tutorial.py [hunk 24]
```diff
@@ -38,258 +51,304 @@
...
+# 2.1. Helper functions
+# ~~~~~~~~~~~~~~~~~~~~~
+#
+
+
+def plot_spectrogram(stft, title="Spectrogram", xlim=None):
+    magnitude = stft.abs()
+    spectrogram = 20 * torch.log10(magnitude + 1e-8).numpy()
+    figure, axis = plt.subplots(1, 1)
+    img = axis.imshow(spectrogram, cmap="nipy_spectral", vmin=-100, vmax=0, origin="lower", aspect="auto")
```
Comment (mthrok) [inline examples/tutorials/mvdr_tutorial.py:84]: The custom colormap is cool, but can we use the default `viridis`?
It is designed so that it is easy for people with color blindness to understand the plot.

See the story behind of the development https://bids.github.io/colormap/

**Review by carolineechen — APPROVED** [review]
